### PR TITLE
Fix edge cases for renderer that could result in map not rendered

### PR DESCRIPTION
**Pull request by kiryldz**
### Summary of changes

Fix several edge-case situations:
 - when new Android surface arrived and previous one is not valid anymore - we additionally recreate native renderer. Previously just EGL core was recreated while native renderer could still reference to outdated GPU resources.
 - when `EGL_CONTEXT_LOST` is triggered during buffer swap (which should not even happen due to other checks) - we re-create EGL as well as native renderer so that it does not reference to outdated GPU resources.

## Pull request checklist:
 - [ ] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
    
 - [ ] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Optimize code for java consumption (`@JvmOverloads`, `@file:JvmName`, etc).
 - [ ] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [ ] Run `make update-api` to update generated api files, if there's public API changes, otherwise the `verify-api-*` CI steps might fail.
 - [ ] Update [CHANGELOG.md](../CHANGELOG.md) or use the label 'skip changelog', otherwise `check changelog` CI step will fail.
 - [ ] If this PR is a `v10.[version]` release branch fix / enhancement, merge it to `main` firstly and then port to `v10.[version]` release branch.

Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).

### CODE REVIEW

File: sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt [hunk 1]
```diff
@@ -311,7 +311,9 @@ internal class MapboxRenderThread : Choreographer.FrameCallback {
       EGL10.EGL_SUCCESS -> { }
       EGL11.EGL_CONTEXT_LOST -> {
         logW(TAG, "Context lost. Waiting for re-acquire")
-        releaseEgl()
+        // Android surface is still valid so we don't release it
+        // otherwise no more rendering will happen
```
Comment (yunikkk) [inline sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt:315]: Is this actually possible, EGL_CONTEXT_LOST and valid surface? Isn't it just happening due to the way you conduct the test?

Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt]: fixed things a bit, PTAL

File: sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt [hunk 2]
```diff
@@ -320,14 +322,6 @@ internal class MapboxRenderThread : Choreographer.FrameCallback {
     }
   }
 
-  private fun releaseEgl() {
```
Comment (kiryldz) [inline sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt:323]: noting that this function is not safe in general - releasing EGL _without_ recreating native renderer may result in native renderer pointing at outdated resources

File: sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt [hunk 3]
```diff
@@ -337,15 +331,21 @@ internal class MapboxRenderThread : Choreographer.FrameCallback {
...
     nonRenderEventQueue.clear()
     renderCreated = false
-    releaseEgl()
-    surface?.release()
+    releaseEglSurface()
+    if (eglPrepared) {
+      eglCore.release()
+    }
+    eglPrepared = false
+    if (releaseAndroidSurface || surface?.isValid == false) {
```
Comment (alexander-kulikovskii) [inline sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt:346]: what about unit tests?

File: sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt [hunk 4]
```diff
@@ -311,7 +311,9 @@ internal class MapboxRenderThread : Choreographer.FrameCallback {
       EGL10.EGL_SUCCESS -> { }
       EGL11.EGL_CONTEXT_LOST -> {
         logW(TAG, "Context lost. Waiting for re-acquire")
-        releaseEgl()
+        // release all resources but not release Android surface if it's still valid -
```
Comment (yunikkk) [inline sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt:314]: If you decide to keep the logic with surface.isValid - move the comment there

**Review by yunikkk — APPROVED** [review]

File: sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt (hunk 3 above)
Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt]: added them

File: sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt (hunk 4 above)
Comment (kiryldz) [reply sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt]: removed it

File: sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt [hunk 5]
```diff
@@ -141,7 +141,7 @@ internal class MapboxRenderThread : Choreographer.FrameCallback {
     )
   }
 
-  private fun setUpRenderThread(creatingSurface: Boolean): Boolean {
+  internal fun setUpRenderThread(creatingSurface: Boolean): Boolean {
```
Comment (alexander-kulikovskii) [inline sdk/src/main/java/com/mapbox/maps/renderer/MapboxRenderThread.kt:144]:
```suggestion
  private fun setUpRenderThread(creatingSurface: Boolean): Boolean {
```

File: sdk/src/test/java/com/mapbox/maps/renderer/MapboxRenderThreadTest.kt [hunk 6]
```diff
@@ -795,4 +796,51 @@ class MapboxRenderThreadTest {
...
+    initRenderThread()
+    provideValidSurface()
+    val oldSurface = surface
+    // simulate Android providing new surface
+    surface = mockk(relaxUnitFun = true)
+    provideValidSurface()
+    // EGL and native renderer are recreated
+    verifyOrder {
+      mapboxRenderer.destroyRenderer()
+      eglCore.release()
```
Comment (alexander-kulikovskii) [inline sdk/src/test/java/com/mapbox/maps/renderer/MapboxRenderThreadTest.kt:840]: +eglCore.releaseSurface(any())

**Review by alexander-kulikovskii — APPROVED** [review]

File: sdk/src/test/java/com/mapbox/maps/renderer/MapboxRenderThreadTest.kt [hunk 7]
```diff
@@ -795,4 +796,53 @@ class MapboxRenderThreadTest {
...
+    idleHandler()
+    every { eglCore.swapBuffers(any()) } returns EGL11.EGL_SUCCESS
+    mapboxRenderThread.queueRenderEvent(MapboxRenderer.repaintRenderEvent)
+    idleHandler()
+    // one swap buffer for surface creation, one for EGL_CONTEXT_LOST, one for EGL_SUCCESS
+    verify(exactly = 3) {
+      eglCore.swapBuffers(any())
+    }
+    // EGL and native renderer are recreated
+    verifyOrder {
```
Comment (yunikkk) [inline sdk/src/test/java/com/mapbox/maps/renderer/MapboxRenderThreadTest.kt:816]: Check that recreation happens right after swap buffers EGL_CONTEXT_LOST?

Comment (kiryldz) [reply sdk/src/test/java/com/mapbox/maps/renderer/MapboxRenderThreadTest.kt:811]: done

**Pull request merged**